### PR TITLE
fix: [ANDROAPP-4952] Improve performance for big option sets

### PR DESCRIPTION
**Pull request by Balcan**
## Description
Please include a summary of the change and include the related jira issue if it exists.

[ jira issue ](https://jira.dhis2.org/browse/ANDROAPP-4952)

## Solution description
If this PR is a fix include a brief description on how the issue is solved.
## Covered unit test cases
Describe the tests that you ran to verify your changes.
## Where did you test this issue?
- [ ] Smartphone Emulator
- [ ] Tablet Emulator
- [ ] Smartphone
- [ ] Tablet
## Which Android versions did you test this issue?
- [ ] 4.4
- [ ] 5.X - 6.X
- [ ] 7.X
- [ ] 8.X
- [ ] 9.X - 10.X
- [ ] Other
## Checklist
- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code

### CODE REVIEW

File: form/src/main/java/org/dhis2/form/data/DataEntryBaseRepository.kt [hunk 1]
```diff
@@ -40,8 +41,20 @@ abstract class DataEntryBaseRepository(
...
+                            optionsToHide = listOf(optionsToHide, optionsInGroupsToHide).flatten(),
+                            optionsToShow = optionsInGroupsToShow
+                        )
+                        is OptionSetConfiguration.DefaultOptionSet -> optionsetConfig.copy(
+                            optionsToHide = listOf(optionsToHide, optionsInGroupsToHide).flatten(),
+                            optionsToShow = optionsInGroupsToShow
+                        )
+                        null -> {
+                            optionSetConfiguration
+                        }
```
Comment (ferdyrod) [inline form/src/main/java/org/dhis2/form/data/DataEntryBaseRepository.kt:56]: If both optionSetConfiguration are doing the same, maybe a simple `if (optionsetConfig != null) { } else { }` would be enough to not duplicate code.

File: form/src/main/java/org/dhis2/form/data/DataEntryBaseRepository.kt [hunk 2]
```diff
@@ -40,8 +41,20 @@ abstract class DataEntryBaseRepository(
...
+                    val optionsetConfig = optionSetConfiguration
+                    this.optionSetConfiguration = when (optionsetConfig) {
+                        is OptionSetConfiguration.BigOptionSet -> optionsetConfig.copy(
+                            optionsToHide = listOf(optionsToHide, optionsInGroupsToHide).flatten(),
+                            optionsToShow = optionsInGroupsToShow
+                        )
+                        is OptionSetConfiguration.DefaultOptionSet -> optionsetConfig.copy(
+                            optionsToHide = listOf(optionsToHide, optionsInGroupsToHide).flatten(),
+                            optionsToShow = optionsInGroupsToShow
+                        )
```
Comment (andresmr) [inline form/src/main/java/org/dhis2/form/data/DataEntryBaseRepository.kt:53]: haven't these two branches the same return object?. It looks like you differentiate between optionsetConfig != null or not. Maybe you could do something like:
`this.optionSetConfiguration = optionsetConfig?.let {
optionsetConfig.copy(
                            optionsToHide = listOf(optionsToHide, optionsInGroupsToHide).flatten(),
                            optionsToShow = optionsInGroupsToShow
                        )
} ?: optionSetConfiguration`

File: form/src/main/java/org/dhis2/form/model/FieldUiModel.kt [hunk 3]
```diff
@@ -55,11 +55,7 @@ interface FieldUiModel {
 
     val renderingType: UiRenderType?
 
-    val options: List<Option>?
-
-    var optionsToHide: List<String>?
-
-    var optionsToShow: List<String>?
+    var optionSetConfiguration: OptionSetConfiguration?
```
Comment (andresmr) [inline form/src/main/java/org/dhis2/form/model/FieldUiModel.kt:58]: very good!

File: form/src/main/java/org/dhis2/form/data/DataEntryBaseRepository.kt (hunk 1 above)
Comment (Balcan) [reply form/src/main/java/org/dhis2/form/data/DataEntryBaseRepository.kt]: Updated to a simpler methor inside the OptionSetConfigation sealed class.

**Review by ferdyrod — APPROVED** [review]

**Review by mmmateos — APPROVED** [review]

**Pull request merged**